File: src/main/java/gargoyle/xenox/util/res/cache/LoadCache.java
package gargoyle.xenox.util.res.cache;

import gargoyle.xenox.util.res.load.Loaders;

import java.net.URL;

public final class LoadCache {
    public static final LoadCache GLOBAL = new LoadCache();
    private final Cache cache = new Cache();

    public <R> R get(Class<R> type, URL location) {
        R resource = cache.get(location.toExternalForm());
        if (resource == null) {
            resource = Loaders.tryLoad(type, location);
            if (resource != null) {
                cache.put(location.toExternalForm(), resource);
            }
        }
        return resource;
    }
}

File: src/main/kotlin/gargoyle/xenox/util/res/cache/LoadCache.kt
package gargoyle.xenox.util.res.cache

import gargoyle.xenox.util.res.load.Loaders
import java.net.URL

class LoadCache {
    private val cache = Cache()
    operator fun <R> get(type: Class<R>, location: URL): R? {
        var resource = cache.get<R>(location.toExternalForm())
        if (resource == null) {
            resource = Loaders.tryLoad(type, location)
            if (resource != null) {
                cache.put(location.toExternalForm(), resource)
            }
        }
        return resource
    }

    companion object {
        val GLOBAL = LoadCache()
    }
}
